File: 2024/src/day02.py
#!/usr/bin/env python3

import os


def relpath(path: str) -> str:
    return os.path.join(os.path.dirname(__file__), path)


reports = []
with open(relpath("../input/day02.txt")) as f:
    for line in f.readlines():
        report = [int(i) for i in line.strip().split()]
        reports.append(report)

Report = list[int]


def sign(x: int) -> int:
    if x > 0:
        return +1
    if x < 0:
        return -1
    return 0


def is_safe(report: Report) -> bool:
    return first_problem(report) is None


def first_problem(report: Report) -> int | None:
    direction = sign(report[1] - report[0])
    for i, (prev, next) in enumerate(zip(report[:-1], report[1:])):
        diff = next - prev
        if direction != sign(diff) or abs(diff) < 1 or abs(diff) > 3:
            return i


def is_safe_tolerant(report: Report) -> bool:
    if is_safe(report):
        return True

    for i in range(len(report)):
        if is_safe(report[:i] + report[i+1:]):
            return True

    return False


n = 0
m = 0
for report in reports:
    if is_safe(report):
        n += 1
    if is_safe_tolerant(report):
        m += 1

print(n)
print(m)

File: 2024/src/day02.zig
const std = @import("std");
const aoc = @import("aoc.zig");

pub fn main() !void {
    var gpa: std.heap.GeneralPurposeAllocator(.{}) = .{};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const file = try std.fs.cwd().openFile("input/day02.txt", .{});
    defer file.close();

    const stdout_file = std.io.getStdOut().writer();
    var bw = std.io.bufferedWriter(stdout_file);
    const stdout = bw.writer();

    var lines = aoc.Lines.init(allocator, file);
    defer lines.deinit();

    var reports = std.ArrayList(Report).init(allocator);
    defer reports.deinit();

    while (try lines.next()) |line| {
        const report = try Report.parse(allocator, line);
        try reports.append(report);
    }

    try stdout.print("{d}\n", .{part1(reports)});
    try bw.flush();

    for (reports.items) |r| {
        r.deinit();
    }
}

const Report = struct {
    levels: std.ArrayList(u32),

    fn parse(allocator: std.mem.Allocator, line: std.ArrayList(u8)) !Report {
        var levels = std.ArrayList(u32).init(allocator);

        var it = std.mem.tokenizeScalar(u8, line.items, ' ');
        while (it.next()) |level| {
            const i = try std.fmt.parseInt(u32, level, 10);
            try levels.append(i);
        }

        return .{ .levels = levels };
    }

    fn deinit(self: Report) void {
        self.levels.deinit();
    }

    fn is_safe(self: Report) bool {
        var prev = self.levels.items[0];

        const sign = self.levels.items[1] > self.levels.items[0];

        var i: usize = 1;
        while (i < self.levels.items.len) {
            const next = self.levels.items[i];
            const diff = abs_diff(next, prev);

            if ((next > prev) != sign) {
                return false;
            }

            if (diff < 1 or diff > 3) {
                return false;
            }

            prev = next;
            i += 1;
        }

        return true;
    }
};

fn abs_diff(a: u32, b: u32) u32 {
    if (a > b) {
        return a - b;
    } else {
        return b - a;
    }
}

fn part1(reports: std.ArrayList(Report)) u32 {
    var n: u32 = 0;
    for (reports.items) |report| {
        if (report.is_safe()) {
            n += 1;
        }
    }

    return n;
}
